### PR TITLE
Remap id's returned in XXXFilteredDimensionSpec.getRow() as per reduced cardinality

### PR DESCRIPTION
**Pull request by himanshug**
fixes the issue described in # #2255

### CODE REVIEW

**Comment by cheddar** [comment]
:+1:

**Comment by binlijin** [comment]
```
public class TopNQueryEngine
{
  public Sequence<Result<TopNResultValue>> query(final TopNQuery query, final StorageAdapter adapter)
  {
   ...
    final Function<Cursor, Result<TopNResultValue>> mapFn = getMapFn(query, adapter);

    return Sequences.filter(
        Sequences.map(
            adapter.makeCursors(filter, queryIntervals.get(0), granularity),
            new Function<Cursor, Result<TopNResultValue>>()
            {
              @Override
              public Result<TopNResultValue> apply(Cursor input)
              {
                log.debug("Running over cursor[%s]", adapter.getInterval(), input.getTime());
                return mapFn.apply(input);
              }
            }
        ),
        Predicates.<Result<TopNResultValue>>notNull()
    );
  }

  private Function<Cursor, Result<TopNResultValue>> getMapFn(TopNQuery query, final StorageAdapter adapter)
  {
    final Capabilities capabilities = adapter.getCapabilities();
    final String dimension = query.getDimensionSpec().getDimension();

   // before dimension decorate
    final int cardinality = adapter.getDimensionCardinality(dimension); 
    ...
    final TopNAlgorithmSelector selector = new TopNAlgorithmSelector(cardinality, numBytesPerRecord);

    final TopNAlgorithm topNAlgorithm;
    ...
    return new TopNMapFn(query, topNAlgorithm);
  }
}
```

**Comment by himanshug** [comment]
@binlijin that cardinality is only use to choose an algorithm, buffer I'm talking about is at https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/query/topn/PooledTopNAlgorithm.java#L67

**Comment by binlijin** [comment]
@himanshug  thanks, i am just want to make sure there is no problem, because i am not familiar with all logic.

File: processing/src/main/java/io/druid/query/dimension/ListFilteredDimensionSpec.java [hunk 1]
```diff
@@ -58,7 +56,7 @@ public ListFilteredDimensionSpec(
   }
```
Comment (guobingkun) [inline processing/src/main/java/io/druid/query/dimension/ListFilteredDimensionSpec.java]: line55 could be: 
```java 
this.isWhitelist = isWhitelist == null || isWhitelist.booleanValue();
```

**Comment by guobingkun** [comment]
:+1:

**Pull request merged**